File: Heimdall/src/pl/edu/agh/heimdall/output/Printer.java
package pl.edu.agh.heimdall.output;

import java.util.HashMap;
import java.util.Map;

import pl.edu.agh.heimdall.consumer.EventSink;
import pl.edu.agh.heimdall.consumer.QueueConsumer;
import pl.edu.agh.heimdall.consumer.QueueEventSink;
import pl.edu.agh.heimdall.events.Event;

public class Printer extends QueueConsumer {
    
    private final Map<String, EventSink> sinks = new HashMap<>();
    
    public Printer(QueueEventSink queue, int bufferSize, int delay) {
        super(queue, bufferSize, delay);
    }
    
    private EventSink getSink(String thread) {
        EventSink sink = sinks.get(thread);
        if (sink == null) {
            sink = new StdoutPrinter(thread);
            sinks.put(thread, sink);
        }
        return sink;
    }

    @Override
    protected void consume(Event e) {
        EventSink sink = getSink(e.thread);
        sink.push(e);
    }

}

File: Gjallarhorn/src/pl/edu/agh/heimdall/gjallarhorn/Printer.scala
package pl.edu.agh.heimdall.gjallarhorn

import pl.edu.agh.heimdall.consumer.EventConsumer
import pl.edu.agh.heimdall.events.Call
import pl.edu.agh.heimdall.events.Throw
import pl.edu.agh.heimdall.events.Return
import pl.edu.agh.heimdall.events.Catch
import pl.edu.agh.heimdall.output.StdoutPrinter
import scala.collection.mutable.Map
import pl.edu.agh.heimdall.consumer.EventSink
import pl.edu.agh.heimdall.events.Event


class Printer extends MultiplexingSink[StdoutPrinter] {

  def newSink(thread: String): StdoutPrinter = new StdoutPrinter(thread)
  
}